core-java-modules/core-java-io-5/src/test/java/com/baeldung/charstreams/CharStreamsUnitTest.java

```diff
@@ -0,0 +1,117 @@
+package com.baeldung.charstreams;

```

Comment:
```suggestion
package com.baeldung.charstreams;

import static org.junit.jupiter.api.Assertions.*;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.nio.file.Paths;
import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThat;
import org.junit.jupiter.api.*;

public class CharStreamsUnitTest {

    private static final DailyTodo DAILY_TODO = new DailyTodo("Code", LocalDateTime.now(), true);
    private static final String EXAMPLE_TEXT = "example";

    private static final String TODO_LIST_PW_FILE = "src/test/resources/pw_todos.txt";
    private static final String TODO_LIST_FW_FILE = "src/test/resources/fw_todos.txt";
    

    @BeforeEach
    public void createFileDirectories() throws IOException {
        Files.createDirectories(Paths.get(TODO_LIST_PW_FILE).getParent());
        Files.createDirectories(Paths.get(TODO_LIST_FW_FILE).getParent());
    }

    @Test
    public void whenUsingTextData_thenPrintWriterWritesToFile() throws IOException {
        try (PrintWriter printWriter = new PrintWriter(new File(TODO_LIST_PW_FILE))) {
            printWriter.print(EXAMPLE_TEXT);
            printWriter.flush();
            assertThat(new File(TODO_LIST_PW_FILE)).hasContent(EXAMPLE_TEXT);
        }
    }

    @Test
    public void whenUsingTextData_thenFileWriterWritesToFile() throws IOException {
        try (FileWriter fileWriter = new FileWriter(new File(TODO_LIST_FW_FILE))) {
            fileWriter.write(EXAMPLE_TEXT);
            fileWriter.flush();
            assertThat(new File(TODO_LIST_FW_FILE)).hasContent(EXAMPLE_TEXT);
        }
    }

    @Test
    public void whenUsingFormattedData_thenPrintWriterWritesToFile() throws IOException {
        try (PrintWriter printWriter = new PrintWriter(new File(TODO_LIST_PW_FILE), StandardCharsets.UTF_8)) {
            printWriter.print(DAILY_TODO);
            printWriter.flush();
            assertThat(new File(TODO_LIST_PW_FILE)).hasContent(DAILY_TODO.toString());
        }
    }

    @Test
    public void whenUsingFormattedData_thenFileWriterWritesToFile() throws IOException {
        try (FileWriter fileWriter = new FileWriter(new File(TODO_LIST_FW_FILE))) {
            fileWriter.write(DAILY_TODO.toString());
            fileWriter.flush();
            assertThat(new File(TODO_LIST_FW_FILE)).hasContent(DAILY_TODO.toString());
        }
    }

    @Test
    public void whenWritingToExistingFile_thenPrintWriterAppendsToFile() throws IOException {
        String existingText = "existing text";
        Files.write(Paths.get(TODO_LIST_PW_FILE), existingText.getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE_NEW);

        try (PrintWriter printWriter = new PrintWriter(new FileOutputStream(TODO_LIST_PW_FILE, true), false, StandardCharsets.UTF_8)) {
            printWriter.append(DAILY_TODO.toString());
            printWriter.flush();
            assertThat(new File(TODO_LIST_PW_FILE)).hasContent(existingText + DAILY_TODO.toString());
        }
    }

    @Test
    public void whenWritingToExistingFile_thenFileWriterAppendsToFile() throws IOException {
        String existingText = "existing text";
        Files.write(Paths.get(TODO_LIST_FW_FILE), existingText.getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE_NEW);

        try (FileWriter fileWriter = new FileWriter(new File(TODO_LIST_FW_FILE), StandardCharsets.UTF_8, true)) {
            fileWriter.append(DAILY_TODO.toString());
            fileWriter.flush();
            assertThat(new File(TODO_LIST_FW_FILE)).hasContent(existingText + DAILY_TODO.toString());
        }
    }

    @Test
    public void whenCheckingError_thenPrintWriterReturnsFalse() throws IOException {
        try (PrintWriter printWriter = new PrintWriter(new File(TODO_LIST_PW_FILE), StandardCharsets.UTF_8)) {
            printWriter.print(DAILY_TODO);
            printWriter.flush();
            assertFalse(printWriter.checkError());
        }
    }

    @Test
    public void whenHandlingError_thenFileWriterHandlesError() {
        try (FileWriter fileWriter = new FileWriter(TODO_LIST_FW_FILE)) {
            fileWriter.write(DAILY_TODO.toString());
        } catch (IOException e) {
            fail("error not expected", e);
        }
    }

    @Test
    public void whenAutoFlushEqualTrue_thenPrintWriterFlushesAutomatically() throws FileNotFoundException {
        try (PrintWriter printWriter = new PrintWriter(new FileOutputStream(TODO_LIST_PW_FILE), true)) {
            printWriter.println(DAILY_TODO);
            assertThat(new File(TODO_LIST_PW_FILE)).hasContent(DAILY_TODO.toString() + System.lineSeparator());
        }
    }
    
    @AfterEach
    public void clean() throws IOException {
        Files.deleteIfExists(Paths.get(TODO_LIST_PW_FILE));
        Files.deleteIfExists(Paths.get(TODO_LIST_FW_FILE));
    }
}
```

@nadiahylary I spent a few minutes to show you what I would do if I was writing these tests

Feel free to keep or adjust this as you wish

Reply:
Thank you.
6 on 7 tests are passing locally. The one faililng is the:
`@Test
    public void whenWritingToExistingFile_thenPrintWriterAppendsToFile() throws IOException {
        String existingText = "existing text";
        Files.write(Paths.get(TODO_LIST_PW_FILE), existingText.getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE_NEW);

        try (PrintWriter printWriter = new PrintWriter(new File(TODO_LIST_PW_FILE), StandardCharsets.UTF_8)) {
            printWriter.append(TODO.toString());
        }

        assertThat(new File(TODO_LIST_PW_FILE)).hasContent(existingText + TODO.toString());
    }`

Reply:
Yes, this is the bug that I told you about here:
https://github.com/eugenp/tutorials/pull/14840#discussion_r1343180479

The bug is caused by the PrintWriter not being constructed in append mode. To fix the problem, you can do this:
```
PrintWriter printWriter = new PrintWriter(new FileOutputStream(TODO_LIST_PW_FILE, true), false, StandardCharsets.UTF_8);
```

Note the `FileOutputStream` which has a `true` append parameter.

Reply:
I fixed it but now it is given a different error: _FileAlreadyExistsException_

Reply:
This is probably because you have a leftover file from a previous partial test run. The simplest solution would be to move the cleanup logic into the `@BeforeEach` and remove the `@AfterEach`

```
   @BeforeEach
    public void createFiles() throws IOException {
        Files.createDirectories(Paths.get(TODO_LIST_PW_FILE).getParent());
        Files.deleteIfExists(Paths.get(TODO_LIST_PW_FILE));
        Files.createDirectories(Paths.get(TODO_LIST_FW_FILE).getParent());
        Files.deleteIfExists(Paths.get(TODO_LIST_FW_FILE));
    }
```

Reply:
if I create a new file and instantly delete it what's the point of creating it at all? And this will generate FileNotFoundException at every test because tge file is deleted immediately after creation.

Reply:
First of all, the `@BeforeEach` suggested does not create any files. It ensures the parent directory of the expected file path is created, so that the actual tests are able to make a strong assumption about that location existing. There is no file creation here, just a parent directory creation.

Secondly, the purpose of deleting in the  `@BeforeEach`, is to cope with cases like the one you may have encountered where, during development, the filesystem is left in an intermediate state (e.g. a test fails during a debugging session and no cleanup code was run). By deleting before the test runs, we are enforcing a precondition for the test, allowing it to assume that the file does not exist at the point in time the test is executed.

Reply:
Okay, everything now works.